src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java

```diff
@@ -0,0 +1,197 @@
+        this.cacheType = Objects.requireNonNull(cacheType);
+        this.cacheNotificationListener = Objects.requireNonNull(cacheNotificationListener);
+        this.vertx = Objects.requireNonNull(vertx);
+        this.metrics = Objects.requireNonNull(metrics);
+        this.clock = Objects.requireNonNull(clock);
+        this.lastResult = new AtomicReference<>();
+    }
+
+    private static Set<String> getInvalidatedKeys(Map<String, String> newMap, Map<String, String> oldMap) {
+        return SetUtils.difference(newMap.keySet(), oldMap.keySet());
+    }
+
+    @Override
+    public void initialize(Promise<Void> initializePromise) {
+        getAll();
+        if (refreshPeriod > 0) {
+            vertx.setPeriodic(refreshPeriod, ignored -> refresh());
+        }
+
+        initializePromise.tryComplete();
+    }
+
+    private void getAll() {
+        final long startTime = clock.millis();
+
+        getFileContentsForDirectory(storedRequestsDirectory)
+                .compose(storedIdToRequest -> getFileContentsForDirectory(storedImpressionsDirectory)
+                        .map(storedIdToImp ->
+                                StoredDataResult.of(storedIdToRequest, storedIdToImp, Collections.emptyList())))
+                .onSuccess(storedDataResult -> handleResult(storedDataResult, startTime, MetricName.initialize))
+                .onFailure(exception -> handleFailure(exception, startTime, MetricName.initialize));
+    }
+
+    private void refresh() {
+        final long startTime = clock.millis();
+
+        getFileContentsForDirectory(storedRequestsDirectory)
+                .compose(storedIdToRequest -> getFileContentsForDirectory(storedImpressionsDirectory)
+                        .map(storedIdToImp ->
+                                StoredDataResult.of(storedIdToRequest, storedIdToImp, Collections.emptyList())))
+                .onSuccess(storedDataResult -> handleResult(invalidate(storedDataResult), startTime, MetricName.update))
+                .onFailure(exception -> handleFailure(exception, startTime, MetricName.update));
+    }
+
+    private Void handleResult(StoredDataResult storedDataResult,
+                              long startTime,
+                              MetricName refreshType) {
+
+        lastResult.set(storedDataResult);
+
+        cacheNotificationListener.save(storedDataResult.getStoredIdToRequest(), storedDataResult.getStoredIdToImp());
+
+        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
+
+        return null;
+    }
+
+    private Future<Void> handleFailure(Throwable exception, long startTime, MetricName refreshType) {
+        logger.warn("Error occurred while request to s3 refresh service", exception);
+
+        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
+        metrics.updateSettingsCacheRefreshErrorMetric(cacheType, refreshType);
+
+        return Future.failedFuture(exception);
+    }
```

Comment:
Let's refactor it a bit (also there is a fix to properly propagate initial success/failure):
```java
@Override
    public void initialize(Promise<Void> initializePromise) {
        final long startTime = clock.millis();

        fetchStoredDataResult()
                .onSuccess(storedDataResult -> handleResult(storedDataResult, startTime, MetricName.initialize))
                .onFailure(exception -> handleFailure(exception, startTime, MetricName.initialize))
                .<Void>mapEmpty()
                .onComplete(initializePromise);

        if (refreshPeriod > 0) {
            vertx.setPeriodic(refreshPeriod, ignored -> refresh());
        }
    }

    private Future<StoredDataResult> fetchStoredDataResult() {
        return Future.all(
                        getFileContentsForDirectory(storedRequestsDirectory),
                        getFileContentsForDirectory(storedImpressionsDirectory))
                .map(CompositeFuture::<Map<String, String>>list)
                .map(results -> StoredDataResult.of(results.getFirst(), results.get(1), Collections.emptyList()));
    }

    private void refresh() {
        final long startTime = clock.millis();

        fetchStoredDataResult()
                .onSuccess(storedDataResult -> handleResult(invalidate(storedDataResult), startTime, MetricName.update))
                .onFailure(exception -> handleFailure(exception, startTime, MetricName.update));
    }

    private void handleResult(StoredDataResult storedDataResult, long startTime, MetricName refreshType) {
        lastResult.set(storedDataResult);

        cacheNotificationListener.save(storedDataResult.getStoredIdToRequest(), storedDataResult.getStoredIdToImp());

        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
    }

    private void handleFailure(Throwable exception, long startTime, MetricName refreshType) {
        logger.warn("Error occurred while request to s3 refresh service", exception);

        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
        metrics.updateSettingsCacheRefreshErrorMetric(cacheType, refreshType);
    }
```